File: bot/bot.py
import logging
import random
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import Application, CommandHandler, MessageHandler, filters, ContextTypes
import sqlite3
from database import db
import requests
import json

# Настройка логирования
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Токен бота
BOT_TOKEN = "8355095598:AAGi48QWU-4e66ZTR2qMYU6aiK-Py1TxjWU"

# Карты Таро
TAROT_CARDS = {
    "Старшие Арканы": [
        "🃏 Шут", "🧙‍♂️ Маг", "🔮 Жрица", "👑 Императрица", "🏛️ Император",
        "⛪ Иерофант", "💑 Влюбленные", "🐎 Колесница", "💪 Сила", "🧘‍♂️ Отшельник",
        "🔄 Колесо Фортуны", "⚖️ Правосудие", "🙏 Повешенный", "💀 Смерть",
        "😇 Умеренность", "😈 Дьявол", "⚡ Башня", "⭐ Звезда", "🌙 Луна",
        "☀️ Солнце", "🎭 Суд", "🌍 Мир"
    ],
    "Младшие Арканы": [
        "✨ Туз Жезлов", "2️⃣ Двойка Жезлов", "3️⃣ Тройка Жезлов", 
        "4️⃣ Четверка Жезлов", "5️⃣ Пятерка Жезлов", "6️⃣ Шестерка Жезлов",
        "7️⃣ Семерка Жезлов", "8️⃣ Восьмерка Жезлов", "9️⃣ Девятка Жезлов",
        "🔟 Десятка Жезлов", "💂‍♂️ Паж Жезлов", "♊ Рыцарь Жезлов",
        "👸 Королева Жезлов", "🤴 Король Жезлов"
    ]
}

class TarotBot:
    def __init__(self):
        self.app = Application.builder().token(BOT_TOKEN).build()
        self.setup_handlers()
    
    def setup_handlers(self):
        self.app.add_handler(CommandHandler("start", self.start))
        self.app.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, self.handle_message))
    
    async def start(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        user_id = update.effective_user.id
        args = context.args
        
        if args:
            if args[0].startswith('ref_'):
                referrer_id = int(args[0].split('_')[1])
                # Начисляем бонусы обоим пользователям
                db.update_balance(user_id, 10)
                db.update_balance(referrer_id, 10)
                await update.message.reply_text("🎉 Вам начислено 10 бесплатных вопросов за регистрацию по ссылке друга!")
            elif args[0] == "balance_10":
                db.update_balance(user_id, 10)
                await update.message.reply_text("✅ Баланс пополнен на 10 вопросов!")
        
        db.create_user(user_id)
        balance = db.get_user_balance(user_id)
        
        welcome_text = f"""
🌟 *Добро пожаловать, {update.effective_user.first_name}!* 🌟

Меня зовут **Спутник** 🧙‍♂️, и я готов помочь вам заглянуть в мир Таро. 

Я умею:
🔮 Делать расклады на ваши вопросы
💫 Давать советы и подсказки
🌙 Помогать разобраться в сложных ситуациях

Просто напишите свой вопрос, и я проведу расклад!

✨ *Ваш баланс: {balance} вопросов* ✨
        """
        
        keyboard = [
            [InlineKeyboardButton("💳 Купить вопросы", url="https://shkarinadust-cmyk.github.io/tarot-bot-miniapp/")],
            [InlineKeyboardButton("👥 Пригласить друзей", callback_data="invite")]
        ]
        reply_markup = InlineKeyboardMarkup(keyboard)
        
        await update.message.reply_text(welcome_text, parse_mode='Markdown', reply_markup=reply_markup)
    
    def generate_tarot_response(self, question, user_id):
        """Генерация ответа от ИИ для Таро"""
        balance = db.get_user_balance(user_id)
        
        if balance <= 0:
            return "❌ Баланс закончился! Пополните: https://shkarinadust-cmyk.github.io/tarot-bot-miniapp/"
        
        # Выбираем случайные карты для расклада
        spread_type = random.choice(["Расклад на ситуацию", "Расклад на день", "Расклад на отношения"])
        cards = random.sample(TAROT_CARDS["Старшие Арканы"] + TAROT_CARDS["Младшие Арканы"], 3)
        positions = ["Прямое", "Перевернутое", "Прямое"]
        
        # Формируем ответ
        response = f"""
🃏 *Запрос:* {question}
🔮 *Расклад:* {spread_type}

*Выпавшие карты:*
1️⃣ **{cards[0]}** - {positions[0]}
2️⃣ **{cards[1]}** - {positions[1]}  
3️⃣ **{cards[2]}** - {positions[2]}

*Интерпретация:*
Карты показывают, что вас ждут интересные события! {cards[0]} говорит о новых возможностях, {cards[1]} указывает на важные решения, а {cards[2]} символизирует гармонию.

*Совет:* Доверьтесь своей интуиции и будьте открыты к переменам! 🌟

💫 Хотите разобрать ситуацию глубже?
        """
        
        # Обновляем баланс
        db.update_balance(user_id, -1)
        new_balance = db.get_user_balance(user_id)
        response += f"\n🔮 *Осталось вопросов: {new_balance}*"
        
        return response
    
    async def handle_message(self, update: Update, context: ContextTypes.DEFAULT_TYPE):
        user_id = update.effective_user.id
        message_text = update.message.text
        
        # Игнорируем приветствия и простые сообщения
        simple_phrases = ["привет", "здравствуй", "hi", "hello", "start", "/start"]
        if message_text.lower() in simple_phrases:
            balance = db.get_user_balance(user_id)
            await update.message.reply_text(f"✨ Привет! Готов помочь с раскладом Таро. Ваш баланс: {balance} вопросов")
            return
        
        # Показываем "типинг"
        await context.bot.send_chat_action(chat_id=update.effective_chat.id, action="typing")
        
        # Генерируем ответ
        response = self.generate_tarot_response(message_text, user_id)
        
        await update.message.reply_text(response, parse_mode='Markdown')
    
    def run(self):
        self.app.run_polling()

if __name__ == "__main__":
    bot = TarotBot()
    bot.run()

File: bot/bot.js
const TelegramBot = require('node-telegram-bot-api');
const axios = require('axios');
require('dotenv').config();

const token = process.env.TELEGRAM_BOT_TOKEN;
const bot = new TelegramBot(token, { polling: true });
const DEEPSEEK_API_KEY = process.env.DEEPSEEK_API_KEY;

// Приветственное сообщение
bot.onText(/\/start/, async (msg) => {
  const chatId = msg.chat.id;
  const userId = msg.from.id;
  
  // Проверяем реферальную ссылку
  const args = msg.text.split(' ');
  if (args[1] === 'balance_10') {
    // Сброс баланса на 10
    await axios.post('http://localhost:3000/api/balance/update', {
      userId,
      amount: 10
    });
  } else if (args[1] && args[1].startsWith('ref_')) {
    // Начисление 10 вопросов за реферала
    await axios.post('http://localhost:3000/api/balance/update', {
      userId,
      amount: 10
    });
  }

  const welcomeMsg = `🔮 *Приветствую, ${msg.from.first_name}!*\n\nМеня зовут *Спутник*, я твой мудрый советчик в мире Таро. 🌙\n\nЯ умею:\n• Делать расклады на ваши вопросы\n• Давать советы по картам\n• Помогать разобраться в ситуациях\n\n*Ваш начальный баланс:* 10 вопросов\n\nНапишите свой вопрос, и мы начнем наше путешествие по картам! ✨`;
  
  bot.sendMessage(chatId, welcomeMsg, { parse_mode: 'Markdown' });
});

// Обработка всех сообщений
bot.on('message', async (msg) => {
  if (msg.text.startsWith('/')) return;
  
  const chatId = msg.chat.id;
  const userId = msg.from.id;
  const userText = msg.text.toLowerCase();
  
  // Проверяем баланс
  const balanceRes = await axios.get(`http://localhost:3000/api/balance/${userId}`);
  const balance = balanceRes.data.balance;
  
  // Если есть слова "карта", "таро", "расклад", "гадание" - считаем вопросом
  const isTarotQuestion = /карт|таро|расклад|гадан|предсказ|будущ|завтра|сегод|вчера|ситуац|отношен|работ|деньг|любов|здоров/.test(userText);
  
  if (!isTarotQuestion) {
    bot.sendMessage(chatId, "Привет! 👋 Задайте вопрос, связанный с картами Таро, и я помогу вам с раскладом! ✨");
    return;
  }
  
  if (balance <= 0) {
    bot.sendMessage(chatId, `❌ *Баланс закончился!*\n\nПополните вопросы через приложение:\nhttps://shkarinadust-cmyk.github.io/tarot-bot-miniapps/`, { parse_mode: 'Markdown' });
    return;
  }
  
  // Отправляем "типинг" (печатает...)
  bot.sendChatAction(chatId, 'typing');
  
  // Запрашиваем ответ у AI
  const aiResponse = await getAIResponse(userText, balance);
  
  // Отправляем ответ
  bot.sendMessage(chatId, aiResponse, { parse_mode: 'Markdown' });
  
  // Уменьшаем баланс на 1
  await axios.post('http://localhost:3000/api/balance/update', {
    userId,
    amount: -1
  });
});

// Функция запроса к DeepSeek
async function getAIResponse(question, balance) {
  try {
    const response = await axios.post('https://api.deepseek.com/v1/chat/completions', {
      model: 'deepseek-chat',
      messages: [{
        role: 'system',
        content: `Ты - мудрый таролог "Спутник". Ты используешь классическую колоду Уэйта. Отвечай на русском, используй эмодзи и жирный шрифт. Структура ответа: 1) Понимание вопроса, 2) Расклад (название карты и положение), 3) Толкование, 4) Итоговый совет. Не больше 10 предложений. Текущий баланс пользователя: ${balance} вопросов. В конце всегда пиши "🔮 Осталось вопросов: ${balance-1}"`
      }, {
        role: 'user',
        content: question
      }],
      max_tokens: 500,
      temperature: 0.7
    }, {
      headers: {
        'Authorization': `Bearer ${DEEPSEEK_API_KEY}`,
        'Content-Type': 'application/json'
      }
    });
    
    return response.data.choices[0].message.content;
  } catch (error) {
    console.error('Ошибка DeepSeek:', error);
    return `✨ *Карта: Шут* (прямое положение)\n\nЭта карта говорит о новых начинаниях и доверии к пути. В вашей ситуации важно сохранить легкость и открыться новым возможностям.\n\n*Совет:* Позвольте себе сделать шаг в неизвестность — это может привести к неожиданным радостям!\n\n🔮 Осталось вопросов: ${balance-1}`;
  }
}